### PR TITLE
feat: Add "read-only" mode

### LINKED ISSUE #2010: Create read-only mode
(add catalog "allow-writes" flag and enforce)

### PR DESCRIPTION
**Pull request by vrongmeal**
```
> create external table local_table from local ( location 'testdata/csv/empty_col.csv' );
Table created
> select * from local_table;
┌───────┬──────┬───────┐
│       │ col1 │ col2  │
│    ── │ ──   │ ──    │
│ Int64 │ Utf8 │ Utf8  │
╞═══════╪══════╪═══════╡
│     0 │ a    │ hello │
│     1 │ b    │ world │
└───────┴──────┴───────┘
> insert into local_table values (2, 'c', 'glaredb');
Error: Cannot write to read-only table: local_table
```

### TODO

- [ ] #2027

Fixes #2010

### CODE REVIEW

File: crates/datafusion_ext/src/runtime/table_provider.rs [hunk 1]
```diff
@@ -21,13 +21,15 @@ use super::runtime_group::RuntimeGroupExec;
 pub struct RuntimeAwareTableProvider {
     pub preference: RuntimePreference,
     pub provider: Arc<dyn TableProvider>,
+    pub allow_writes: Option<bool>,
```
Comment (tychoish) [inline crates/datafusion_ext/src/runtime/table_provider.rs:24]: this should definitely be an enum for "allowed_operations" (read, write_ddl, write_dml, write_all) or something like that. 

(there are potentially a few states for this, and the difference between false and not-specified-optional is confusing)

**Review by tychoish — COMMENTED** [review]
This looks good, but I think I'd also:
- [x] enum
- [ ] adding a couple of tests seems important

File: crates/datafusion_ext/src/runtime/table_provider.rs (hunk 1 above)
Comment (vrongmeal) [reply crates/datafusion_ext/src/runtime/table_provider.rs:24]: > this should definitely be an enum for "allowed_operations" (read, write_ddl, write_dml, write_all) or something like that.

Agreed. Will update this and add tests.

File: testdata/sqllogictests/allowed_operations.slt [hunk 2]
```diff
@@ -0,0 +1,14 @@
+statement ok
+CREATE EXTERNAL DATABASE debug_db FROM debug;
```
Comment (tychoish) [inline testdata/sqllogictests/allowed_operations.slt:2]: this is cool, and I definitely 100% got to this point and don't fully understand how this works or would catch a bug if we accidentally regressed?

**Review by tychoish — COMMENTED** [review]
this looks really cool and more or less like what I'd expect, but I definitely would like other folks to look at this too

File: testdata/sqllogictests/allowed_operations.slt (hunk 2 above)
Comment (vrongmeal) [reply testdata/sqllogictests/allowed_operations.slt:2]: We'll catch a bug here. Trying to insert should always check for permissions first. Moreover, we have an error message in the check that should force us to add different tests in the future if something regressed.

**Review by tychoish — APPROVED** [review]

**Pull request merged**